src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs

```diff
@@ -2829,6 +2894,21 @@ public async Task HubMethodInvokeDoesNotCountTowardsClientTimeout()
 
                     await hubMethodTask.OrTimeout();
 
+                    // There is a small window when the hub method finishes and the timer starts again
+                    // So we need to delay a little before ticking the heart beat.
+                    // We do this by waiting until we know the HubConnectionHandler code is in pipe.ReadAsync()
+                    var timeout = DateTime.UtcNow;
+                    var timeoutSeconds = -30;
+                    while (!customDuplex.WrappedPipeReader.IsWaitingForRead() && timeout > DateTime.UtcNow.AddSeconds(timeoutSeconds))
```

Comment:
While the logic seems right, it does seem a little backwards. Why not add 30 seconds to DateTime.UtcNow to get the timeout timestamp and then compare that to DateTime.UtcNow directly? That way you don't have to AddSeconds every iteration.